### PR TITLE
ref(grouptype): Update usages of GroupType

### PR DESCRIPTION
**Pull request by ceorourke**
Replace usages of the `GroupType` enum, `GROUP_TYPE_TO_CATEGORY`, `GROUP_TYPE_TO_TEXT`, `PERFORMANCE_TYPES`, `PROFILE_TYPES`, and `GROUP_CATEGORY_TO_TYPES` mappings with the newer dataclass version(s). 

See https://github.com/getsentry/sentry/issues/43484 for the full context behind this change, but the tl;dr is this makes it simpler to add a new group type rather than needing to update several variables.

### CODE REVIEW

File: src/sentry/event_manager.py [hunk 1]
```diff
@@ -2331,12 +2337,12 @@ def _save_aggregate_performance(jobs: Sequence[PerformanceJob], projects: Projec
                         problem = performance_problems_by_hash[new_grouphash]
 
                         span.set_tag("create_group_transaction.outcome", "no_group")
-                        span.set_tag("group_type", problem.type.name)
+                        span.set_tag("group_type", problem.type.slug)
                         metric_tags["create_group_transaction.outcome"] = "no_group"
-                        metric_tags["group_type"] = problem.type.name
+                        metric_tags["group_type"] = problem.type.slug
```
[the remark this answers is not in the document]
Comment (ceorourke) [reply src/sentry/event_manager.py:2342]: The only difference is the capitalization, e.g. ERROR -> error

File: src/sentry/types/issues.py [hunk 2]
```diff
@@ -1,22 +1,21 @@
...
-    PERFORMANCE_SLOW_DB_QUERY = 1001
-    PERFORMANCE_RENDER_BLOCKING_ASSET_SPAN = 1004
-    PERFORMANCE_N_PLUS_ONE_DB_QUERIES = 1006
-    PERFORMANCE_CONSECUTIVE_DB_QUERIES = 1007
-    PERFORMANCE_FILE_IO_MAIN_THREAD = 1008
-    PERFORMANCE_N_PLUS_ONE_API_CALLS = 1010
-    PERFORMANCE_M_N_PLUS_ONE_DB_QUERIES = 1011
-    PERFORMANCE_UNCOMPRESSED_ASSETS = 1012
-    PROFILE_BLOCKED_THREAD = 2000
+# class GroupType(Enum):
```
[the remark this answers is not in the document]
Comment (ceorourke) [reply src/sentry/types/issues.py:4]: Ah yeah I just left it in while I was working on it for easy reference

File: src/sentry/models/group.py [hunk 3]
```diff
@@ -432,17 +440,17 @@ class Group(Model):
...
             (
-                GroupType.PERFORMANCE_RENDER_BLOCKING_ASSET_SPAN.value,
+                PerformanceRenderBlockingAssetSpanGroupType.type_id,
                 _("Large Render Blocking Asset"),
             ),
             (
-                GroupType.PERFORMANCE_N_PLUS_ONE_API_CALLS.value,
+                PerformanceNPlusOneAPICallsGroupType.type_id,
                 _("N+1 API Calls"),
             ),
```
[the remark this answers is not in the document]
Comment (ceorourke) [reply src/sentry/models/group.py:455]: Is it ok to just remove `choices` entirely?

File: src/sentry/rules/history/preview_strategy.py [hunk 4]
```diff
@@ -16,8 +16,8 @@
 
 # Maps group category to dataset
 GROUP_CATEGORY_TO_DATASET: Dict[GroupCategory, Dataset] = {
-    GroupCategory.ERROR: Dataset.Events,
-    GroupCategory.PERFORMANCE: Dataset.Transactions,
+    GroupCategory.ERROR.value: Dataset.Events,
+    GroupCategory.PERFORMANCE.value: Dataset.Transactions,
```
[the remark this answers is not in the document]
Comment (ceorourke) [reply src/sentry/rules/history/preview_strategy.py:20]: Sure, I don't really know what this is, was just madly updating everything.

File: src/sentry/models/group.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (wedamija) [reply src/sentry/models/group.py:455]: Yeah, that's fine. Django doesn't do any validation here

File: src/sentry/utils/performance_issues/performance_problem.py [hunk 5]
```diff
@@ -1,7 +1,7 @@
 from dataclasses import dataclass
 from typing import Any, Mapping, Optional, Sequence
 
-from sentry.types.issues import GROUP_TYPE_TO_TEXT, GroupType
+from sentry.grouptype.grouptype import GroupType, get_group_type_by_type_id
 
 
 @dataclass
```
[the remark this answers is not in the document]
Comment (ceorourke) [reply src/sentry/utils/performance_issues/performance_problem.py:7]: What do you mean?

File: src/sentry/event_manager.py [hunk 6]
```diff
@@ -156,12 +163,11 @@
...
 GROUPHASH_IGNORE_LIMIT_MAP = {
-    GroupType.PERFORMANCE_N_PLUS_ONE_DB_QUERIES: 3,
-    GroupType.PERFORMANCE_SLOW_DB_QUERY: 100,
-    GroupType.PERFORMANCE_CONSECUTIVE_DB_QUERIES: 15,
-    GroupType.PERFORMANCE_UNCOMPRESSED_ASSETS: 100,
+    PerformanceNPlusOneGroupType: PerformanceNPlusOneGroupType.ignore_limit,
+    PerformanceSlowDBQueryGroupType: PerformanceSlowDBQueryGroupType.ignore_limit,
+    PerformanceConsecutiveDBQueriesGroupType: PerformanceConsecutiveDBQueriesGroupType.ignore_limit,
+    PerformanceUncompressedAssetsGroupType: PerformanceUncompressedAssetsGroupType.ignore_limit,
 }
```
Comment (wedamija) [inline src/sentry/event_manager.py:171]: Is this map used anymore?

File: src/sentry/grouptype/grouptype.py [hunk 7]
```diff
@@ -11,31 +15,42 @@ class GroupType:
...
+
+
+def get_group_type_by_slug(slug: str) -> Any:
+    if slug not in _slug_lookup.keys():
+        raise ValueError(f"No group type with the slug {slug} is registered.")
+    return _slug_lookup[slug]
+
+
+def get_group_type_by_type_id(id: int) -> Any:
+    if id not in _group_type_registry.keys():
```
Comment (wedamija) [inline src/sentry/grouptype/grouptype.py]: Nit: `.keys()` isn't necessary

File: src/sentry/utils/performance_issues/performance_problem.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wedamija) [reply src/sentry/utils/performance_issues/performance_problem.py:7]: I think I confused myself, nm

File: src/sentry/event_manager.py (hunk 6 above)
Comment (ceorourke) [reply src/sentry/event_manager.py:171]: Oops no, it kept changing during rebases and I forgot to remove it.

**Review by wedamija — APPROVED** [review]
LGTM. This is reasonably risky just since it's touching a lot of files, so let's follow the deploy steps here to isolate/monitor the deploy in case anything breaks.

https://www.notion.so/sentry/RFC-Ensuring-Safe-Deployments-39ab017e4e914e2a9fa427a456302e0a

**Pull request merged**